app/src/main/java/de/westnordost/streetcomplete/quests/surface/AddRoadSurfaceForm.kt

```diff
@@ -13,18 +14,38 @@ class AddRoadSurfaceForm : AImageListQuestAnswerFragment<Surface, SurfaceAnswer>
 
     override fun onClickOk(selectedItems: List<Surface>) {
```

Comment:
I find the data flow here quite confusing. Passing `removeTracktype` to a function that primarily asks for surface description, or a function with `and` in its name that does two different things look like code smells.

Can it be refactored to something like this? Then each function has one single clear purpose:

```kt
override fun onClick(selectedItems: List<Surface>) {
    val surface = selectedItems.single()
    confirmTracktypeMismatch(surface) { shouldRemoveTracktype ->
        collectSurfaceDescription(surface) { description ->
            applyAnswer(value, description, shouldRemoveTracktype)
        }
    }
}

private fun confirmTracktypeMismatch(
    surface: Surface,
    onTracktypeConfirmed: (shouldRemoveTracktype: Boolean) -> Unit
) {
    val tracktype = osmElement!!.tags["tracktype"]
    if (tracktype == null) {
        onTracktypeConfirmed(false)
    } else if (isSurfaceAndTracktypeMismatching(surface.osmValue, tracktype)) {
        AlertDialog.Builder(requireContext())
            .setTitle(R.string.quest_generic_confirmation_title)
            .setMessage(R.string.quest_surface_tractypeMismatchInput_confirmation_description)
            .setPositiveButton(R.string.quest_generic_confirmation_yes) { _, _ ->
                onTracktypeConfirmed(true)
            }
            .setNegativeButton(R.string.quest_generic_confirmation_no, null)
            .show()
    }
}

private fun collectSurfaceDescription(
    surface: Surface,
    onSurfaceDescribed: (description: String?) -> Unit
) {
    if (!surface.shouldBeDescribed) {
        onSurfaceDescribed(null)
    } else {
        AlertDialog.Builder(requireContext())
            .setMessage(R.string.quest_surface_detailed_answer_impossible_confirmation)
            .setPositiveButton(R.string.quest_generic_confirmation_yes) { _, _ ->
                DescribeGenericSurfaceDialog(requireContext(), onSurfaceDescribed).show()
            }
            .setNegativeButton(android.R.string.cancel, null)
            .show()
    }
}
```

Reply:
I admit that any time I see lambda expression I need to stop for minute and think about it, but original code is hairy and smelly.

`confirmTracktypeMismatch` - what you think about `confirmPotentialTracktypeMismatch` or `investigateTracktypeMismatch`?

`confirmTracktypeMismatch` sounds a bit weird to me as confirmation will be needed only in rare cases

Reply:
`confirmPotentialTracktypeMismatch` sounds good

Reply:
used with a minor improvements